drivers: uhc: implement UHC shim driver for NXP KHCI and EHCI controller

MarkWangChinese opened the pull request:
Enablet the NXP MCUX UHC drivers.
Add uhc support for frdm_k22f, rt1060, lpc55s69 and lpc55s28. frdm_k22f supports the KHCI controller, rt1060 supports the EHCI controller, lpc55s69 and lpc55s28 support the OHCI and IP3516HS controller.

MarkWangChinese committed:
drivers: uhc: implement nxp mcux uhc driver

It is based on SDK USB Host controller driver.
Support NXP EHCI, KHCI, OHCI and IP3516HS controllers.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

MarkWangChinese committed:
boards: nxp: add uhc support for frdm_k22f, rt1060, lpc55s69 and lpc55s28

add uhc related items to dts.
add clock initialization
add BM4 if CONFIG_USB_UHC_NXP_KHCI is enabled
add pin mux
update board related CMakeLists.txt
update sdk-ng CMake to include NXP controller drivers
update west.yml to contain the hal_nxp pr

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

PetervdPerk-NXP commented:
Any plans to enable 1170 and 1180 as well?
Since they share almost the same EHCI controller.

tmon-nordic commented on include/zephyr/usb/usbh.h:
```diff
@@ -137,6 +137,24 @@ int usbh_disable(struct usbh_contex *uhs_ctx);
+ *
+ * @return the device's address.
+ */
+uint8_t usbh_udev_get_address(void *usb_device);
```
Why `void *usb_device`? Shouldn't it simply be `struct usb_device *usb_device`?

MarkWangChinese replied:
It is called from context that can't see `struct usb_device`. For example: the uhc files that can't see `struct usb_device`.

tmon-nordic replied:
Just treat it as opaque pointer then. Having `void *` here is bad.

MarkWangChinese replied:
In the `struct uhc_transfer`, it is `void *udev`. So I use the `void *`. Could you please give reference codes how to treat it as opaque pointer?

tmon-nordic replied:
That would need a change in the `struct uhc_transfer` then. Opaque pointers in C are simply achieved by having incomplete struct definition. So in this case it would be `struct usb_device;` above `struct uhc_transfer {` and then the `void *udev;` can be changed to `struct usb_device *udev;`. Similar could be done with the transfer completion callback. I somehow completely missed this during the uhc review.

MarkWangChinese replied:
Thanks @tmon-nordic, I changed it, please help to review again.

tmon-nordic commented on include/zephyr/usb/usbh.h:
```diff
@@ -89,6 +89,9 @@ struct usbh_class_data {
 	int (*resumed)(struct usbh_contex *const uhs_ctx);
 };
 
+/* declare usb_device */
```
This comment doesn't really make anything clearer, I'd remove it.

jfischer-no commented on drivers/usb/uhc/uhc_mcux_ehci.c:
```diff
@@ -0,0 +1,540 @@
+					 mcux_ep_handle);
+		}
+
+		if ((mcux_xfer->setupPacket->bmRequestType & USB_REQUEST_TYPE_DIR_MASK) ==
```
:crystal_ball: I see a future where all these shim drivers fail because they pull their own USB_* definitions.

jfischer-no commented on drivers/usb/uhc/uhc_mcux_ip3516hs.c:
```diff
@@ -0,0 +1,492 @@
+{
+	const struct uhc_mcux_config *config = dev->config;
+
+	/* disable interrupt. */
+	config->irq_disable_func(dev);
+	return 0;
+}
+
+static int uhc_mcux_shutdown(const struct device *dev)
+{
+	const struct uhc_mcux_config *config = dev->config;
+	struct uhc_mcux_data *priv = uhc_get_private(dev);
+
+	/* disable interrupt. */
+	config->irq_disable_func(dev);
+	/* disable MCUX USB host drive task */
```
Please cleanup.

jfischer-no commented on drivers/usb/uhc/uhc_mcux_common.c:
```diff
@@ -0,0 +1,133 @@
+usb_status_t USB_HostHelperGetPeripheralInformation(usb_device_handle deviceHandle,
+						    uint32_t infoCode, uint32_t *infoValue)
+{
+	const struct device *dev = usbh_udev_get_dev(deviceHandle);
```
Driver must not call anything from the higher layer. Each uhc_transfer has a device address, you have to use it somehow.

MarkWangChinese replied:
The purpose of I add `usbh_udev_get_dev` and `usbh_udev_get_address` is to get device's information by controller driver. I think the device's related hub information (hub address, port num, first hs hub address, first hs hub port num) and speed are needed by the controller driver to do the transfer. Will you plan to put all the information to every `uhc_transfer`?

jfischer-no replied:
Regardless of whether this information is needed or not, you cannot include this header or use any higher-level API in the driver.

MarkWangChinese replied:
Even the linux EHCI driver include the usb.h `linux/usb.h>` in linux/drivers/usb/host/ehci-hcd.c. Actually I don't understand why the simple helper API can't be called in lower level. Just want to figure out the style, do you mean that only callback is right from lower level to higher level? Anyway I have changed the codes to get device information based on the xfer. When developing the usb host stack, I think the codes may need to be updated to match host stack's design. For now, it should work.

jfischer-no replied:
> Actually I don't understand why the simple helper API can't be called in lower level.

Because it is always a level violation, and it ends in a nightmare.

>I think the device's related hub information (hub address, port num, first hs hub address, first hs hub port num) and speed are also needed by the controller driver to do the transfer. Will you plan to put all the information to every uhc_transfer?

Yes, but maybe rather a pointer to a device structure similar to `struct urb`. struct usb_device would have to be defined in uhc.h and replace addr in `struct uhc_transfer`.

MarkWangChinese replied:
OK, Thanks for the explanation.

MarkWangChinese commented on drivers/usb/uhc/Kconfig.mcux:
```diff
@@ -0,0 +1,46 @@
+	bool "NXP MCUX USB EHCI Host controller driver"
+	default y
+	depends on DT_HAS_NXP_UHC_EHCI_ENABLED
+	select EVENTS
```
The NXP hal controller drivers use the k_event. Hal controller drivers depend on the event of modules\hal\nxp\mcux\mcux-sdk\components\osa\fsl_os_abstraction_zephyr, and the fsl_os_abstraction_zephyr depend on k_event.

MarkWangChinese replied:
The MCUX usb host hal controller driver is like the UHC of Zephyr. In the hal driver, one task function is provided, the upper layer can use the task function to create one task to handle the controller interrupts. Then the OS event is used to handle interrupts in this task function.

jfischer-no replied:
HAL driver should not use Zephyr events, otherwise it is not a HAL driver.

MarkWangChinese commented on drivers/usb/uhc/uhc_mcux_ehci.c:
```diff
@@ -0,0 +1,540 @@
+	const struct pinctrl_dev_config *pincfg;
+};
+
+#if defined(CONFIG_NOCACHE_MEMORY)
+/* allocate non-cached buffer for usb */
+static void *uhc_mcux_nocache_alloc(uint32_t size)
+{
+	void *p = (void *)k_heap_alloc(&mcux_transfer_alloc_pool, size, K_NO_WAIT);
+
+	if (p != NULL) {
+		(void)memset(p, 0, size);
+	}
+
+	return p;
+}
+
+/* free the allocated non-cached buffer */
+static void uhc_mcux_nocache_free(void *p)
+{
+	k_heap_free(&mcux_transfer_alloc_pool, p);
+}
+#endif
```
Which platform and cmd that you are using? Give me the cmd line that you use, I will check building errors.

jfischer-no replied:
cmake -DBOARD=mimxrt1050_evk -DCONF_FILE=device_and_host_prj.conf zephyr/samples/subsys/usb/shell/

Fails with:
```
CMake Error at /home/jofi/zephyr-project/zephyr/cmake/modules/extensions.cmake:1944 (add_subdirectory):
  The binary directory

    /home/jofi/zephyr-project/zephyr/build/modules/hal_nxp/usb

  is already used to build a source directory.  It cannot be used to build
  source directory

    /home/jofi/zephyr-project/zephyr/modules/hal_nxp/usb

  Specify a unique binary directory name.
Call Stack (most recent call first):
  /home/jofi/zephyr-project/zephyr/modules/hal_nxp/CMakeLists.txt:12 (add_subdirectory_ifdef)
```

jfischer-no replied:
Please add 

```
zephyr_uhc0: &usb1 {
	status = "okay";
};
```

to `boards/nxp/mimxrt1050_evk/mimxrt1050_evk.dts`

MarkWangChinese replied:
I updated the `boards/nxp/mimxrt1050_evk/mimxrt1050_evk.dts`, I add both `zephyr_uhc0 `and `zephyr_uhc1`. The `zephyr_uhc0` and `zephyr_udc0` use same usb HW IP controller, so only one (device function or host function) can be initialized and enabled. If the usb controller 0 is used by `zephyr_udc0`, then `zephyr_uhc0` can't be used to enable the same controller as host function.
I do the follow test.
`west build -b mimxrt1050_evk .\samples\subsys\usb\shell -- -DCONF_FILE="device_and_host_prj.conf"`
run the binary and attach one usb flash drive to the usb port. input the follow shell cmds:
```
usbh int
usbh enable
usbh device address 1
usbh device descriptor device
```

jfischer-no replied:
I noticed that the second controller on mimxrt1050 does not generate SoF.

MarkWangChinese commented on drivers/usb/uhc/uhc_mcux_ehci.c:
```diff
@@ -0,0 +1,540 @@
+	return 0;
+}
+
+static int uhc_mcux_dequeue(const struct device *dev, struct uhc_transfer *const xfer)
+{
+	/* TODO */
+	return 0;
+}
```
OK

jfischer-no replied:
dequeue must be implemented, e.g. USB/IP relies on it.
https://github.com/zephyrproject-rtos/zephyr/blob/e9eee4eb232b674aaa0f63f892cb291883abfedf/drivers/usb/uhc/uhc_virtual.c#L572

CrushedB commented:
> > Any plans to enable 1170 and 1180 as well? Since they share almost the same EHCI controller.
> 
> After this basic NXP UHC controller driver merges, I think next step is implementing the host stack. Because the host stack is not ready, so supporting more platforms does not make much sense.

Hi @MarkWangChinese,
Thanks for this delivery.
Can you please elaborate on the missing host stack? What won't be supported after this delivery? (I would like to be able to host a USB thumb drive on the rx10xx board).

DerekSnell commented:
Hi @CrushedB ,
> Can you please elaborate on the missing host stack? What won't be supported after this delivery? (I would like to be able to host a USB thumb drive on the rx10xx board).

Currently, Zephyr does not include a USB Host stack, only USB Device.  This [USB RFC](https://github.com/zephyrproject-rtos/zephyr/issues/42066) tracks the plans and progress for adding a USB Host stack.  Today, the USB Host Controller (UHC) driver API has be introduced for the low-level APIs.  And this PR is porting the UHC APIs to the NXP USB hardware.  But as the RFC shows, the USB Host stack for the high-level APIs still needs to be introduced.  And the RFC shows the plans for the Host classes that will be supported.

If you have any further questions on the USB Host plans or progress, we encourage you to engage in the [USB RFC](https://github.com/zephyrproject-rtos/zephyr/issues/42066).  Best regards

MarkWangChinese commented:
Rebase the Zephyr main branch.
Hi @jfischer-no Could you please help to review again?
I have another query, I see you have one task item as "Introduction of USB host stack implementation (WIP @jfischer-no)" in https://github.com/zephyrproject-rtos/zephyr/issues/42066. When will you plan to give this introducation?

tmon-nordic commented:
Twister failures sounds like some hal issue with missing `fsl_os_abstraction.h` header. Please fix it.

MarkWangChinese commented:
> Twister failures sounds like some hal issue with missing `fsl_os_abstraction.h` header. Please fix it.

Fixed

jfischer-no commented on drivers/usb/uhc/uhc_mcux_common.c:
```diff
@@ -0,0 +1,252 @@
+	info_code = (usb_host_dev_info_t)infoCode;
+	switch (info_code) {
+	case kUSB_HostGetDeviceAddress:
+		*infoValue = (uint8_t)xfer->addr;
```
addr has benn removed in favor of `struct usb_device *udev`
`		*infoValue = xfer->udev->addr;`

MarkWangChinese replied:
updated.

jfischer-no commented on drivers/usb/uhc/uhc_mcux_common.c:
```diff
@@ -0,0 +1,252 @@
+		/* TODO: workaround. current stack doesn't support to
+		 * get spped from controller driver.
+		 */
+		*infoValue = UHC_EVT_DEV_CONNECTED_HS;
```
Please use `xfer->udev->speed`

MarkWangChinese replied:
Updated.

jfischer-no commented on drivers/usb/uhc/uhc_mcux_common.c:
```diff
@@ -0,0 +1,252 @@
+	uint8_t i;
+
+	/* if already initialized */
+	mcux_ep_handle = uhc_mcux_get_hal_ep(priv, xfer->udev, xfer->ep);
```
The driver should also take xfer->mps into account, as this can change depending on the interface alternate, or to handle device reset correctly, as mps0 is unknown at startup and set to default 8.

MarkWangChinese replied:
Updated.
Will you plan to add the similar controller interface like the `ep_enable` and `ep_disable` of `struct udc_api`? 
I think the host controller driver needs to maintain some information/resource (for example: split transaction) about the `udev + ep`, then the controller interface is needed to alloc/free the information in controller driver.
In the follow cases, I think the endpoint information/resource needs be freed: (1) The attached device is detached with/without hub; (2) The alternate setting is switched, the previous interface's eps need to be freed and initialize the new eps based on the alternate setting interface; (3) The class driver stop to use one interface of the composite device, the interface related eps needs to be freed.
Without the controller interface, controller driver doesn't know when to free the ep resource.

jfischer-no commented on drivers/usb/uhc/uhc_mcux_common.c:
```diff
@@ -0,0 +1,252 @@
+	 * 'number per uframe' and the endpoint type cannot be got yet.
+	 */
+	pipe_init.numberPerUframe = 0; /* TODO: need right way to implement it. */
+	pipe_init.interval = 0;        /* TODO: need right way to implement it. */
```
xfer->interval (Currently only used by USBIP)

MarkWangChinese replied:
updated.

jfischer-no commented on drivers/usb/uhc/uhc_mcux_ehci.c:
```diff
@@ -0,0 +1,427 @@
+
+	ret = uhc_mcux_bus_control(dev, kUSB_HostBusReset);
+	if (ret == 0) {
+		ret = uhc_mcux_bus_control(dev, kUSB_HostBusRestart);
```
This should not be necessary, it will end up in an endless loop of resets because it generates a new UHC_EVT_DEV_CONNECTED_* event. (Testing your PR, I also found a bug in usbh_shell.c, will fix it ASAP.)

MarkWangChinese replied:
Updated.

jfischer-no commented on drivers/usb/uhc/uhc_mcux_ehci.c:
```diff
@@ -0,0 +1,427 @@
+
+			/* set address to let USB_HostHelperGetPeripheralInformation get right value
+			 */
+			xfer->addr = (uint8_t)transfer->setupPacket->wValue;
```
xfer->udev->addr, but driver must not write to xfer->udev->addr.

MarkWangChinese replied:
Updated.

jfischer-no reviewed (changes requested):
@MarkWangChinese Please rebase

MarkWangChinese commented:
Hi @jfischer-no In the NXP host controller driver, there is one task function that process the interrupt of the controller and the transfers by using the `Event`, mcux uhc uses the task function to create task. mcux uhc codes don't use the `Event` directly. If you think the NXP host controller driver can't use `EVENT` because it is HAL driver, could we treat NXP host controller driver as non-hal driver?

MarkWangChinese committed:
boards: nxp: add uhc support for mimxrt1050_evk

add uhc related items to dts.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

jfischer-no commented:
@MarkWangChinese Do you want to wait for https://github.com/zephyrproject-rtos/zephyr/pull/86920 or get this PR in and remove the bounce buffers afterwards?

MarkWangChinese committed:
drivers: uhc: udc: use DYNAMIC_INTERRUPTS for KHCI and EHCI

Because the same IP supports both device and host, so use
DYNAMIC_INTERRUPTS for KHCI and EHCI if both device and host
are enabled.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

jfischer-no commented:
@MarkWangChinese Please fix style issues reported by compliance checks.

MarkWangChinese commented:
> @MarkWangChinese Please fix style issues reported by compliance checks.

Fixed.

harristomy commented:
@dleach02 @MarkWangChinese any updates on https://github.com/zephyrproject-rtos/hal_nxp/pull/496 ? It'd be great if we had this feature in Zephyr 4.2

matt-wood-ct commented:
Can we have board support for the RT1040? 
That's the part I'm using and are keen to test on as classes start to arrive :smile:

dleach02 commented:
Sorry @jfischer-no @tmon-nordic, I needed to rebase to pull in a hot fix from main that was causing CI to fail here. Need your +1 reapplied so we can merge
